keps/sig-api-machinery/2887-openapi-enum-types/README.md

```diff
@@ -0,0 +1,836 @@
+Protocol can be shown as a dropdown of possible values instead of a textbox.
+The user no longer needs to see an error of invalid value after submits the request.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Add a marker that indicates a field as an enum.
+- Make the Kubernetes OpenAPI generator recognize the marker.
+- Deduce values of an enum type from its Go type definition.
+- Detect and annotate enum types in all built-in types. 
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+This KEP does not require removal of existing validation of enum types. While OpenAPI-based
+of enum types can simplify this type of validation, retiring custom-made validations is out
+of scope.
+
+This proposal does not intend to change the enum syntax for third-party tools like `kubebuilder`.
+The implementation should only suggest changes to internal APIs. 
+
+## Proposal
+
+When generating OpenAPI schema, the de facto enum types should be reflected as enums in the resulting schema.
+
+We define enum types with following properties:
+- All enum types are closed, i.e., the type includes a finite number of possible values.
+- All enum types should be string. There can be no enum integers, floats, etc
+
+Additionally, This KEP makes the following assumptions on the implementation:
+- All enum types have their own type alias. i.e. , no raw string enums.
+- The canonical definition of a type is written in go and follows the standard 
+  Kubernetes convention.
+
+### Enum Pattern
+
+The Enum Pattern of a type, appearing in a go source file, is defined as follows:
+```go
+package internal
+
+// FooType define sa type alias, underlying type must be string
+type FooType string
+
+// in the same package, define constants with the alias
+const (
+    BarConst FooType = "Bar"
+    BazConst FooType = "Baz"
+)
+
+// in one or more resources or nested objects, one of the field is of 
+// the defined alias type.
+
+// ExampleResource should refer FooType
+type ExampleResource struct {
+    FooField FooType
+}
+
+```
+Enum Type is the type that follows such pattern (`FooType` in the example above).
+Enum Values are the possible values of the Enum Type (`Bar` and `Baz` are Enum Values for `FooType`).
+
+### Enum Tag
+The Enum Tag is the marker `// +enum` that must be in the closest comment blocks
+of an enum type following Enum Pattern.
+
+```go
+package internal
+
+// FooType define sa type alias, underlying type must be string
+// +enum
+type FooType string
+```
+
+An actual example looks like
+```go
+package core
+
+// Protocol defines network protocols supported for things like container ports.
+// +enum
+type Protocol string
+```
+
+### Detect Possible Enum Types
+
+A new API rule will be added to `kube-openapi` that warns about detected Enum Pattern
```

Comment:
Why would we do this? I'm not sure I see the benefit, especially for alpha/beta.

Reply:
For post-beta, this linter would warn if the user forget an enum marker. This should be added when it gets close to GA.

Reply:
I still don't see the benefit if we know that there will be false positives anyway.

Reply:
@apelisse This is the "linter" that detects existing enums during its first run.

Reply:
False positives are very rare. I've been run a tons of experiments and only find one counter-example. The vest majority are real enums. The linter can prevent the user from forgetting to add the marker.

Reply:
I suggest we keep the linter for now. If it proves to be little value for maintainers, we can remove it before GA.

Reply:
omitted for now. we can re-design this part later.